### PR TITLE
feat(gatsby): add DANGEROUSLY_DISABLE_OOM

### PR DESCRIPTION
**Pull request by wardpeet**
## Description
This adds a process.env variable called `DANGEROUSLY_DISABLE_OOM` to disable saving the redux state to disk. This helps with going out of memory in the bootstrap phase. The downside is that your build will be much slower as it will always build from a clean dir.

We won't put this variable inside our docs as it's a feature we don't want to highlight or people to use.

## Related Issues
#12485

### CODE REVIEW

File: packages/gatsby/src/redux/index.js [hunk 1]
```diff
@@ -73,6 +74,16 @@ const store = Redux.createStore(
 
 // Persist state.
 function saveState() {
+  if (process.env.DANGEROUSLY_DISABLE_OOM) {
+    // meausure how many people are using DANGEROUSLY_DISABLE_OOM in build statements
+    telemetry.decorateEvent(`BUILD_END`, {
```
Comment (wardpeet) [inline packages/gatsby/src/redux/index.js]: @jamo wdyt of this? Maybe I should just omit it 😛

Comment (jamo) [reply packages/gatsby/src/redux/index.js]: I like this a lot! 
We(I) need to allow this key in the API before we can ship it.

Comment (wardpeet) [reply packages/gatsby/src/redux/index.js]: Maybe it's better to omit this for now? So the PR isn't blocked by this?
cc @jamo

**Comment by KyleAMathews** [comment]
Is this intended as a temporary hack? E.g. once v8.seralize stuff gets figured out we remove this?

**Comment by muescha** [comment]
i would add it anyhow to documentation with an big warning to avoid more and more "hidden preferences" or avoid some surprising behavior.

anyhow it would name it somehow what it do and not what it avoid `DANGEROUSLY_DISABLE_SAVING_REDUX_STATE`

**Comment by pieh** [comment]
Yes it's meant to be temporary. Users are blocked on OOM issues and until `v8.serialize` land, there is nothing they can do to unblock themselves (other than monkey patching and commenting out `JSON.stringify` parts in `saveState` function.

@muescha This is intentionally not mentioning redux state. We don't want to encourage using it for the purpose of skipping redux state caching (as then we can never fix cache issues if users don't use it at all). This is only to unblock cases of out of memory issues. This will also be removed once proper fix is merged.

**Comment by wardpeet** [comment]
Yes temporary yes. Let me just remove the analytics and get this merged.

File: packages/gatsby/src/redux/index.js (hunk 1 above)
Comment (jamo) [reply packages/gatsby/src/redux/index.js]: sorry for the delays and a good call.

**Comment by muescha** [comment]
maybe comment in code why its temporary and what it solves ... just to see it later in code reviews and don't forget to remove it later

**Comment by DSchau** [comment]
@wardpeet I'm gonna pick up this PR and add some documentation if that's cool!

**Pull request merged**